notebook/notebookapp.py

```diff
@@ -1820,10 +1854,17 @@ def launch_browser(self):
         else:
             open_file = self.browser_open_file
 
-        b = lambda: browser.open(
-            urljoin('file:', pathname2url(open_file)),
-            new=self.webbrowser_open_new)
-        threading.Thread(target=b).start()
+        if not self.use_redirect_file:
+            if browser:
+                b = lambda: browser.open(url_path_join(self.connection_url, uri),
+                                        new=self.webbrowser_open_new)
+                threading.Thread(target=b).start()
+
+        else:
+            b = lambda: browser.open(
+                urljoin('file:', pathname2url(open_file)),
+                new=self.webbrowser_open_new)
+            threading.Thread(target=b).start()
```

Comment:
The `browser.open()` and the thread call can be shared between the code paths - only assembling the URL needs to be different.